### PR TITLE
add differentiable SmallPT geometry demo

### PR DESCRIPTION
**Pull request by a-elrawy**
This PR introduces new demos showcasing Clad's automatic differentiation applied to 3D computer graphics problems.

The demos utilize soft rasterization (smooth proximity weight per pixel-to-sphere distance) to create a differentiable rendering pipeline. This allows `clad::gradient` to flow continuous gradients from image pixels directly to scene parameters via a Mean Squared Error (MSE) loss function.

Included Demos:

- **Camera Pose Optimization** (`SmallPT_DiffGeom.cpp`): Recovers an unknown camera position from a target image via gradient descent.

- **Object Position Optimization** (`SmallPT_DiffSphere.cpp`): Recovers the 3D coordinates of a scene object (sphere) by matching its silhouette to a target image.

- **Light Source Recovery** (`SmallPT_DiffLight.cpp`): Recovers an unknown light source's position by differentiating a shading model, demonstrating Differentiable Shading for realistic scene reconstruction

**Visualization Tools:**
To help analyze the convergence behavior, a Python utility `generate_static.py` is included [here](https://github.com/user-attachments/files/26219583/generate_static.py). This script processes the optimization data saved by the demos to generate summary plots.

**How to use:**

Run any of the demos to generate visualization data in `demos/ComputerGraphics/smallpt/visualizations/`.
Execute `python3 generate_static.py` to produce a .png summary for each demo.

### CODE REVIEW

File: test/Misc/RunDemos.C [hunk 1]
```diff
@@ -4,6 +4,12 @@
 // RUN: %cladclang %S/../../demos/DebuggingClad.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/RosenbrockFunction.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/ComputerGraphics/smallpt/SmallPT.cpp -I%S/../../include 2>&1
+// RUN: %cladclang %S/../../demos/ComputerGraphics/smallpt/SmallPT_DiffGeom.cpp -I%S/../../include -oSmallPT_DiffGeom.out 2>&1
+// RUN: ./SmallPT_DiffGeom.out | FileCheck -check-prefix CHECK_SMALLPT_DIFFGEOM %s
+// CHECK_SMALLPT_DIFFGEOM: Converged
+// RUN: %cladclang %S/../../demos/ComputerGraphics/smallpt/SmallPT_DiffSphere.cpp -I%S/../../include -oSmallPT_DiffSphere.out 2>&1
+// RUN: ./SmallPT_DiffSphere.out | FileCheck -check-prefix CHECK_SMALLPT_DIFFSPHERE %s
+// CHECK_SMALLPT_DIFFSPHERE: Converged
```
Comment (vgvassilev) [inline test/Misc/RunDemos.C:12]: How long does it take to execute this demo?

Comment (a-elrawy) [reply test/Misc/RunDemos.C:12]: both are about 10ms

`time /tmp/SmallPT_DiffSphere.out` :
```
real    0m0.010s
user    0m0.006s
sys     0m0.000s
```

`time /tmp/SmallPT_DiffGeom.out` :
```
real    0m0.010s
user    0m0.003s
sys     0m0.000s
```

I'm currently working on another one for light recovery
`time /tmp/SmallPT_DiffLight` :
```
real    0m0.010s
user    0m0.004s
sys     0m0.000s
```

**Review by alexander-penev — APPROVED** [review]
OK, but it would be good to do a refactoring in the future to base all tests on the code in SmallPT.cpp‎
